btree/btree.c

```diff
@@ -3740,6 +3767,58 @@ int64_t btree_open_tick(struct m0_sm_op *smop)
 	}
 }
 
+/**
+ * btree_close_tree_tick function is used to traverse through different states
+ * to facilitate the working of m0_btree_close().
+ *
+ * @param smop     represents the state machine operation
+ * @return int64_t returns the next state to be executed.
+ */
+int64_t btree_close_tree_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop     = M0_AMB(bop, smop, bo_op);
+	struct m0_btree       *tree    = bop->bo_arbor;
+	struct td             *td_curr = tree->t_desc;
+	struct nd             *nd_head = ndlist_tlist_head(&td_curr->
+							   t_active_nds);
+	struct nd             *nd_curr;
+
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if (td_curr->t_ref > 1)
+			tree_put(td_curr);
+		else if (td_curr->t_ref == 1) {
+			nd_curr = nd_head;
+			if (td_curr->t_starttime == 0)
+				td_curr->t_starttime = m0_time_now();
+
+			if (ndlist_tlist_length(&td_curr->t_active_nds) > 1)
+				return P_TIMECHECK;
```

Comment:
This is not what I meant when I said to have a state for time check. This code is same as old code just time check moved to a state. I had asked because there is a chance that the "time check" part will be repeated over and over again. Hence you want to come to this state directly, instead of again going through  all the above steps. 
If you want the same code, you can put back your code as it was earlier and remove P_TIMECHECK, as there is no difference between this and the last code. It just adds one more cycle of going out and coming back in the state machine for P_TIMECHECK.

Reply:
Oh Sorry for the misinterpretation.

Reply:
I will try it this way.

Reply:
Updated